### ISSUE TITLE
TEG exponential moving average needs to be replaced with a better solution

### ISSUE THREAD

**Issue by ArtisticRoomba** [space-wizards/space-station-14#37902]
## Description
Right now the TEG's power output is smoothed using an exponential moving average. This magics an insane amount of power to the grid and isn't a very good solution.

This was brought on by https://github.com/space-wizards/space-station-14/pull/37658, we merged this as a band-aid as it was causing large problems (triggering epilepsy).

It was discussed to replace this with a large thermal buffer to prevent large jumps while still keeping power output realistic to what it should be. However any solution that ensures that power isn't being magiced out of thin air while still smoothing the TEG in pressure fluttering situations is good.

**Media**
See the pull request media.

**Comment by ArtisticRoomba** [comment]
So what I'm probably going to do as a naive first try:
- Add an imaginary configurable thermal mass to each circulator (clearly there's a one ton copper block hidden under the hood of each circulator)
- Gas from the pipes is transferred to that thermal mass at a certain efficiency (slowly but surely)
- Have the efficiency rating extract our desired % from those thermal masses

It'll take some tweaking as we just decreased the efficiency to make extracting energy from low specific heat gasses (oxygen and nitrogen) very hard on purpose. So if we increase the efficiency when going from pipe to buffer, we'll just be back to square one when it comes to killing single loops.

My only problem with the above solution is that the TEG naturally wants to produce and dump as much power humanly possible to the grid. This will always result in the TEG draining the thermal mass of energy before the peaks and valleys of a fluttering TEG can warm the thermal buffer. So we're proxying the heat only to follow the same behavior.

This is probably tangentially related, but we could enforce manual TEG throttling using TEG temperature ranges. Have the TEG consume thermal energy from the heat buffers as if it was matching to the grid via power ramping. However, if the circulators got too hot (from atmosians dumping too much heat into the system) it would overheat and be damaged. This would accomplish two things:
- Prevent atmosians from running everything pedal to the metal, a goal of the atmos roadmap
- Encourage atmosians to monitor their TEG and preform grid matching, another goal of the atmos roadmap (major setup and forget solutions shouldn't be a thing)

Let me know what you think of my analysis but that's something I had concern about when thinking about it a bit further.

**Comment by ArtisticRoomba** [comment]
I will re-implement the TEG throttling logic which was removed in https://github.com/space-wizards/space-station-14/pull/29050/.

Game plan is:
- Gas transfers 10% of heat to some imaginary thermal mass (necessary to keep single-loops dead).
- This thermal mass will allow massive peaks of heat energy from a fluttering TEG to be dumped while still allowing power flow to be smooth.
- The amount of heat required to generate enough power to fulfill the supply, considering the current thermal efficiency (the actual $n$) of the TEG, is transferred between the circulators.

This will allow us to nuke the EMA.

**Comment by slarticodefast** [comment]
Ok, I would do the following:
- Don't do "10% of the heat" since that violates a whole bunch of physics laws. It would also mean the reservoir would get hotter than the gas.
- Instead bring the gas and the reservoir into thermal contact for a short time
  $dQ = \kappa (T_{gas} - T_{reservoir}) dt$ where $\kappa$ is some conductivity constant.
- Alternatively make it completely equalize temperature with the reservoir $T_{final} = \frac{T_{gas} C_{gas}+T_{reservoir}C_{reservoir}}{C_{gas}+C_{reservoir}}$ but that would have the side effect that a single loop would be enough. You can still use this equation to clamp the heat transfer though.
- $C_{reservoir}$ should be large enough so that a gas with a very high heat capacity (e.g. Frezon) with maximum pressure and flow rate won't instantly change the reservoirs temperature by large amounts.
- The temperature of both reservoirs should be shown in the examination text for better feedback.
